app/config/services.yml

```diff
@@ -1,83 +1,121 @@
+
+    # The other section defines a rule to automatically register and autowire the
+    # controller classes found in the src/AppBundle/Controller/ directory.
+    #
+    # By default all services are made private according to the global "_defaults"
+    # section at the top of this file.
+    #
+    # However, in Symfony, controllers must always be declared public in order to
+    # be lazy instanciated when they're really needed. This is why the inherited
+    # default "public" attribute is overriden to force the registered controller
+    # services to be marked public.
+    AppBundle\Controller\:
+        resource: '../../src/AppBundle/Controller'
+        public: true
+
+    # This third party Twig extension must be manually registered as a service
+    # because its class doesn't live in any of the previous defined directories.
+    #
+    # Thus, its class is not namespaced and cannot be defined in a global rule.
+    # Indeed, registering new classes whose filename matches a glob pattern as
+    # services like in the two previous sections only works for namespaced classes.
     app.twig.intl_extension:
-        public: false
         class:  Twig_Extensions_Extension_Intl
-        tags:
-            - { name: twig.extension }
 
-    # Defining a form type as a service is only required when the form type
-    # needs to use some other services, such as the entity manager.
-    # See http://symfony.com/doc/current/best_practices/forms.html
-    app.form.type.tagsinput:
-        class: AppBundle\Form\Type\TagsInputType
-        arguments: ['@doctrine.orm.entity_manager']
-        tags:
-            - { name: form.type }
+    # Some classes cannot be fully autowired because their methods accept either
+    # some scalar arguments that Symfony cannot guess or a typehinted dependency
+    # for which the container has at least two registered services matching the
+    # type.
+    #
+    # Both the "AppBundle\Twig\AppExtension" and "AppBundle\EventListener\RedirectToPreferredLocaleSubscriber"
+    # classes have a "__construct()" method that receives a "$locales" scalar argument
+    # that Symfony cannot guess. This is why we must manually and explicitly provide
+    # the wiring of this argument to complete their service definitions.
+    #
+    # To do so, the remaining unwired named arguments must be defined with their
+    # corresponding values. Here, the "$locales" named argument is configured to
+    # receive the value of the global "%app_locales%" parameter defined under the
+    # "parameters" section of the "app/config/config.yml" file.
+    #
+    # Note that the order in which the named arguments are defined below doesn't
+    # matter as they're referenced here by their real names in the PHP code.
+    # Symfony is then smart enough to make the corresponding matching when compiling,
+    # optimizing and dumping the container as a raw PHP class in the cache directory.
+    AppBundle\Twig\AppExtension:
```

Comment:
The first section lets Symfony register the classes from these directories as services. It'll make its best to autowire everything. But some service requires some extra manual configuration when Symfony cannot guess the remaining parts (dependency collision or scalar arguments). This is what the third section does.

The second section called `instanceof` allows to append configuration to every definitions that match the rule type. It has nothing to do with autowiring! This is probably the confusing part that must be well taught to newcomers. It's just a way to apply a group of configuration settings to each service definition matching the rule. It's a way to make configuration not repeat itself. This means you can still manually define your services and use the `instanceof` section to append the DI tags or extra method calls to your sets of configured services.